### DISCUSSION TITLE
Unexpected token B in JSON at position 14

### DISCUSSION THREAD

**Discussion by JV300381** [jreyesr/n8n-nodes-carbonejs#12] [Q&A] [answered]
Hi @jreyesr,

I have moved on trying a more complex scenario and am being faced with the issue below while loading the json data into the Carbone node.

Unexpected token B in JSON at position 14

On another note, it seems that Aggregators is not supported? E.g [aggSum](https://carbone.io/documentation.html#aggsum)
(Error: Formatter "aggSum" does not exist. Do you mean "and"?)

More details below:

The workflow:

![issue0](https://github.com/user-attachments/assets/ecc4021e-9305-4578-853b-be45f6a9ef82)

The issue:

![issue1](https://github.com/user-attachments/assets/7ca02fed-2adb-498b-9c47-6f1db224e5ed)

![issue2](https://github.com/user-attachments/assets/2b0eb462-cd3f-4f1a-8aa2-7fecaeb115f6)

Strangely, when I load the json data into the context directly, its working.

![issue3](https://github.com/user-attachments/assets/2298d214-a611-4fc2-beae-3b8eab151c3a)

![issue4](https://github.com/user-attachments/assets/12d1b2f7-d601-410a-a671-a0f4f9bae247)

Below is the full error log from n8n:

{
  "errorMessage": "Unexpected token B in JSON at position 14",
  "errorDetails": {},
  "n8nDetails": {
    "nodeName": "Carbone",
    "nodeType": "n8n-nodes-carbonejs.carboneNode",
    "nodeVersion": 1,
    "operation": "render",
    "time": "9/19/2024, 8:59:17 AM",
    "n8nVersion": "1.58.2 (Self Hosted)",
    "binaryDataMode": "default",
    "cause": {},
    "stackTrace": [
      "NodeOperationError: Unexpected token B in JSON at position 14",
      "    at Object.execute (C:\\Users\\Jessen\\.n8n\\nodes\\node_modules\\n8n-nodes-carbonejs\\nodes\\CarboneNode\\CarboneNode.node.ts:245:12)",
      "    at Workflow.runNode (C:\\Users\\Jessen\\AppData\\Roaming\\npm\\node_modules\\n8n\\node_modules\\n8n-workflow\\src\\Workflow.ts:1383:31)",
      "    at C:\\Users\\Jessen\\AppData\\Roaming\\npm\\node_modules\\n8n\\node_modules\\n8n-core\\src\\WorkflowExecute.ts:1080:42",
      "    at C:\\Users\\Jessen\\AppData\\Roaming\\npm\\node_modules\\n8n\\node_modules\\n8n-core\\src\\WorkflowExecute.ts:1800:11"
    ]
  }
}

Error log from Terminal:

NodeOperationError: Unexpected token B in JSON at position 14
    at Object.execute (C:\Users\Jessen\.n8n\nodes\node_modules\n8n-nodes-carbonejs\nodes\CarboneNode\CarboneNode.node.ts:245:12)
    at Workflow.runNode (C:\Users\Jessen\AppData\Roaming\npm\node_modules\n8n\node_modules\n8n-workflow\src\Workflow.ts:1383:31)
    ... 2 lines matching cause stack trace ...
    at C:\Users\Jessen\AppData\Roaming\npm\node_modules\n8n\node_modules\n8n-core\src\WorkflowExecute.ts:1800:11 {
  cause: SyntaxError: Unexpected token B in JSON at position 14
      at JSON.parse (<anonymous>)
      at Object.execute (C:\Users\Jessen\.n8n\nodes\node_modules\n8n-nodes-carbonejs\nodes\CarboneNode\CarboneNode.node.ts:182:27)
      at Workflow.runNode (C:\Users\Jessen\AppData\Roaming\npm\node_modules\n8n\node_modules\n8n-workflow\src\Workflow.ts:1383:31)
      at C:\Users\Jessen\AppData\Roaming\npm\node_modules\n8n\node_modules\n8n-core\src\WorkflowExecute.ts:1080:42
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
      at C:\Users\Jessen\AppData\Roaming\npm\node_modules\n8n\node_modules\n8n-core\src\WorkflowExecute.ts:1800:11,
  node: {
    parameters: {
      operation: 'render',
      context: '={"reportname":{{ $json.reportname }},"reportvalue":{{ $json.reportvalue }}, "reporttype":{{ $json.reporttype }}}',
      dataPropertyName: '=data',
      dataPropertyNameOut: 'data',
      options: {}
    },
    id: '141dc5da-be59-43c1-89f6-927461f5be7e',
    name: 'Carbone',
    type: 'n8n-nodes-carbonejs.carboneNode',
    typeVersion: 1,
    position: [ 40, 380 ],
    executeOnce: true
  },
  timestamp: 1726723387954
} undefined
SyntaxError: Unexpected token B in JSON at position 14
    at JSON.parse (<anonymous>)
    at Object.execute (C:\Users\Jessen\.n8n\nodes\node_modules\n8n-nodes-carbonejs\nodes\CarboneNode\CarboneNode.node.ts:182:27)
    at Workflow.runNode (C:\Users\Jessen\AppData\Roaming\npm\node_modules\n8n\node_modules\n8n-workflow\src\Workflow.ts:1383:31)
    at C:\Users\Jessen\AppData\Roaming\npm\node_modules\n8n\node_modules\n8n-core\src\WorkflowExecute.ts:1080:42
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at C:\Users\Jessen\AppData\Roaming\npm\node_modules\n8n\node_modules\n8n-core\src\WorkflowExecute.ts:1800:11 undefined

[payslip.docx](https://github.com/user-attachments/files/17053940/payslip.docx)
[payslip.json](https://github.com/user-attachments/files/17053941/payslip.json)

**Comment by jreyesr** [comment, marked as answer]
Hello @JV300381!

I think that the first issue, the Unexpected token B, is because the JSON that is being provided to the Context field isn't actually valid JSON. The values that are strings (namely, BASIC SALARY and EARNINGS) should be wrapped in double quotes:

```json
{"reportname":"BASIC SALARY","reportvalue":35000,"reporttype":"EARNINGS"}
```

Double quotes are missing here:

![image](https://github.com/user-attachments/assets/fc4c0877-9127-4664-b6a5-04d428b33c89)

The JSON that is being passed now, that is, `{"reportname":BASIC SALARY,"reportvalue":35000, "reporttype":EARNINGS}`, doesn't comply with the JSON syntax. After `{"reportname":`, there should come either:
* A string (e.g. `{"reportname":"A STRING", ...}`), which is what you want, and strings always start with a double quote `"`, not B
* A number (e.g. `{"reportname":123, ...}`), which can't start with a B, only with the minus sign - or a digit 0 to 9
* One of the literal words `null`, `true` or `false` (e.g. `{"reportname":true, ...}`)
* The start of a nested document (e.g. `{"reportname":{...}, ...}`), which requires a `{`
* The start of an array (e.g. `{"reportname":[...], ...}`), which requires a `[`

To fix that, try the following Expression on the Context field. Note that quotes have been added around the two `{{ $json.X }}` fields that should be strings. The `reportvalue` field doesn't actually need them, because it can be interpreted correctly as a number.

```
{"reportname":"{{ $json.reportname }}","reportvalue":{{ $json.reportvalue }}, "reporttype":"{{ $json.reporttype }}"
```

> Strangely, when I load the json data into the context directly, its working.

Yes, that's because when you hardcoded the JSON data in the Context, it contained the correct quotes, so it parsed correctly

---

Regarding the missing aggregators: Yes, sadly [Aggregators are an Enterprise/Cloud-only feature](https://carbone.io/documentation.html#aggregators). Carbone Enterprise is paid (though note that there's a free Cloud tier that provides, I think, 100 documents per month at no cost, but requires a credit card) and uses a different method of interaction (it's based on REST API calls, rather than an embedded Javascript library). If you're able to use Enterprise (say, you're doing this for a company and can convince them to cough up [the requisite money](https://carbone.io/pricing.html), or the free Cloud tier is enough for you) and sending documents off to the cloud isn't a deal breaker, you should be able to use [this community node](https://www.npmjs.com/package/n8n-nodes-carbone) instead of mine. Or just use the HTTP Request node, that should work too, since the API [is fairly simple](https://carbone.io/api-reference.html#render-reports).

If you'd rather stick with local rendering, I'd recommend precomputing those aggregates _before_ rendering happens. For example, say that you want to do this (print the sum of the `qty` field of all `cars`):

![image](https://github.com/user-attachments/assets/f9626e2b-7eba-4279-84c8-affde0d70bde)

You could instead add a Set node before the Carbone node and add a new field with the following Expression (or something that does the same, I just happen to find the functional map-then-reduce style natural for this case):

```
{{ $json.cars.map(x => x.qty).sum() }}
```

This first picks out the `qty` field of each element of `cars` (and returns an array like `[1, 4, 3, 2, 1, 10]`, and then the `.sum()` collapses all that into a single number, 21. That Set node would add a `cars_qty_sum` field to the JSON document that would contain the sum of all `qty` fields:

![image](https://github.com/user-attachments/assets/580dab95-8285-4995-9501-856e4d11ed5f)

And then you can pass that JSON document to the template and just do `{d.cars_qty_sum}` instead of `{d.cars[].qty:aggSum}`, which isn't available.
